### ISSUE TITLE
PDF conversions and auto image normalization

### ISSUE THREAD

**Issue by AntJanus** [creatale/node-dv#16]
I was wondering if anyone has tried to tackle PDF conversions built-in or what the proposed solution to the PDF-to-JPG conversion would be. Obviously, a lot of scanners convert straight to PDFs rather than images.

Also, as far as normalizing the image background, straightening the text, etc. is there an option for that anywhere?

**Comment by v4l3r10** [comment]
I tried different solutions for use a pdf in nodeDV.
My best POC has been to use [node-pdfutils](https://github.com/Gottox/node-pdfutils) for reading the pdf, convert every page in a PNG  buffer and, finally, use it in node-dv.

**Comment by WolfgangFellger** [comment]
Adding JPEG natively was a relatively easy choice since the library codebase is already large anyway, but PDF is nontrivial and quite far out of scope. As said above, use another library to convert it intermediately to raw pixel buffers or PNG - I'd not recommend JPEG because of the information loss.

If by 'normalizing' you mean 'stretch image contrast to use the entire color range', this should be doable using histogram() and applyCurve(). If not, please explain in more detail.

Basically, if Leptonica has the operation, we will gladly pass it outside, otherwise it's probably out of scope – this library is mostly meant as a wrapper, we do not wish to implement any non-trivial algorithms in it.

**Comment by AntJanus** [comment]
I ended up rolling my own lib with Ghostscript so I got that (mostly) figured out. I totally understand that it's out of scope. :)

I guess what I meant was adding contrast, straightening the image, and other post processing to help the OCR read the text easier. 

Thanks for the help!

**Comment by AntJanus** [comment]
@WolfgangFellger I'm wondering if you could explain how to do the contrasting (and I'm guessing thresholding) with histogram and applyCurve.

**Comment by WolfgangFellger** [comment]
I'll use CoffeeScript for brevity and only handle grayscale. As an example, this is (roughly) what 'Colors/Auto/Normalize' in GIMP would do:

```coffeescript
min = 255
max = 0

# Find maximum and minimum sample value present in image
for value, index in image.histogram()
	if value > 0 and index < min
		min = index
	if value > 0 and index > max
		max = index

# Map the range [min...max] out to [0..255] by linear interpolation
valueSpread = 255 / (max - min)
mapping = (Math.floor((value - min) * valueSpread) for value in [0..255])

# Apply mapping
image.applyCurve mapping
```
When constructing `mapping`, `min` changes brightness, while `valueSpread` changes contrast  (except image editing software will typically center at 127, not 0 like this). Basically, the input to `applyCurve` is what you would draw in GIMP's Color/Curves.

**Comment by AntJanus** [comment]
Thanks @WolfgangFellger! That's really helpful. I decided to try thresholding (with Otsu's method) and deskewing. I'll try the results and then try what you're suggesting.

:)

**Comment by AntJanus** [comment]
On a sidenote, just to give information to anyone that comes across this issue:

1. Tesseract does take input PDFs; however, it will fail on malformed PDFs so it's a better practice to use something like Ghostscript
2. There's a function called `findSkew` and `rotate`, Find skew tries to find the angle the image is skewed and rotate will rotate it
3. Image normalization via `applyCurve` is above, or you can use `threshold` (but you'd still have to read the `histogram` for it).
